File: sdks/java/basys.sdk/src/main/java/org/eclipse/basyx/aas/backend/connected/aas/submodelelement/ConnectedDataElement.java
package org.eclipse.basyx.aas.backend.connected.aas.submodelelement;

import java.util.HashSet;
import java.util.Set;

import org.eclipse.basyx.aas.api.metamodel.aas.qualifier.qualifiable.IConstraint;
import org.eclipse.basyx.aas.api.metamodel.aas.reference.IReference;
import org.eclipse.basyx.aas.backend.connected.facades.ConnectedHasDataSpecificationFacade;
import org.eclipse.basyx.aas.backend.connected.facades.ConnectedHasKindFacade;
import org.eclipse.basyx.aas.backend.connected.facades.ConnectedHasSemanticsFacade;
import org.eclipse.basyx.aas.backend.connected.facades.ConnectedQualifiableFacade;
import org.eclipse.basyx.aas.backend.connected.facades.ConnectedReferableFacade;
import org.eclipse.basyx.aas.impl.metamodel.hashmap.aas.qualifier.Referable;
import org.eclipse.basyx.vab.core.proxy.VABElementProxy;
/**
 * "Connected" implementation of DataElement
 * @author rajashek
 *
 */
public class ConnectedDataElement extends ConnectedSubmodelElement {
	public ConnectedDataElement(VABElementProxy proxy) {
		super(proxy);		
	}
	
	@Override
	public HashSet<IReference> getDataSpecificationReferences() {
		return new ConnectedHasDataSpecificationFacade(getProxy()).getDataSpecificationReferences();
	}

	@Override
	public void setDataSpecificationReferences(HashSet<IReference> ref) {
		new ConnectedHasDataSpecificationFacade(getProxy()).setDataSpecificationReferences(ref);
		
	}
	
	@Override
	public String getIdshort() {
		return new ConnectedReferableFacade(getProxy()).getIdshort();
	}

	@Override
	public String getCategory() {
		return new ConnectedReferableFacade(getProxy()).getCategory();
	}

	@Override
	public String getDescription() {
		return new ConnectedReferableFacade(getProxy()).getDescription();
	}

	@Override
	public IReference getParent() {
		return new ConnectedReferableFacade(getProxy()).getParent();
	}

	@Override
	public void setIdshort(String idShort) {
		 new ConnectedReferableFacade(getProxy()).setIdshort(idShort);
		
	}

	@Override
	public void setCategory(String category) {
		 new ConnectedReferableFacade(getProxy()).setCategory(category);
		
	}

	@Override
	public void setDescription(String description) {
		 new ConnectedReferableFacade(getProxy()).setDescription(description);
		
	}

	@Override
	public void setParent(IReference obj) {
		 new ConnectedReferableFacade(getProxy()).setParent(obj);
		
	}
	
	@Override
	public void setQualifier(Set<IConstraint> qualifiers) {
		new ConnectedQualifiableFacade(getProxy()).setQualifier(qualifiers);
		
	}

	@Override
	public Set<IConstraint> getQualifier() {
		return new ConnectedQualifiableFacade(getProxy()).getQualifier();
	}
	
	@Override
	public IReference getSemanticId() {
		return new ConnectedHasSemanticsFacade(getProxy()).getSemanticId();
	}

	@Override
	public void setSemanticID(IReference ref) {
		 new ConnectedHasSemanticsFacade(getProxy()).setSemanticID(ref);
		
	}
	
	@Override
	public String getHasKindReference() {
		return  new ConnectedHasKindFacade(getProxy()).getHasKindReference();
	}

	@Override
	public void setHasKindReference(String kind) {
		new ConnectedHasKindFacade(getProxy()).setHasKindReference(kind);
		
	}
	
	@Override
	public String getId() {
		// try local get
		String id = (String) this.getLocal(Referable.IDSHORT);
		if (id != null) {
			return id;
		}
		return (String) getProxy().getModelPropertyValue(Referable.IDSHORT);
	}

	@Override
	public void setId(String id) {
		// try set local if exists
		if (this.getLocal(Referable.IDSHORT) != null) {
			this.putLocal(Referable.IDSHORT, id);
		}
		getProxy().setModelPropertyValue(Referable.IDSHORT, id);
		
	}
}

File: sdks/c++/basys.sdk.cc/src/aas/aas/backend/connected/ConnectedDataElement.cpp
/*
 * ConnectedDataElement.cpp
 *
 *      Author: wendel
 */

#include "ConnectedDataElement.h"
#include "basyx/types.h"

#include <memory>


namespace basyx {
namespace aas {
namespace backend {

ConnectedDataElement::ConnectedDataElement(std::shared_ptr<vab::core::proxy::IVABElementProxy> proxy) :
  ConnectedSubmodelElement(proxy)
{}

basyx::objectCollection_t ConnectedDataElement::getDataSpecificationReferences() const
{
  return this->getProxyCollection(reference::paths::DATASPECIFICATIONS);
}

void ConnectedDataElement::setDataSpecificationReferences(const basyx::objectCollection_t & data_specification_references)
{
  this->setProxyValue(reference::paths::DATASPECIFICATIONS, data_specification_references);
}

std::string ConnectedDataElement::getIdShort() const
{
  return this->getProxyValue(qualifier::ReferablePaths::IDSHORT);
}

void ConnectedDataElement::setIdShort(const std::string & idShort)
{
  this->setProxyValue(qualifier::ReferablePaths::IDSHORT, idShort);
}

std::string ConnectedDataElement::getCategory() const
{
  return this->getProxyValue(qualifier::ReferablePaths::CATEGORY);
}

void ConnectedDataElement::setCategory(const std::string & category)
{
  this->setProxyValue(qualifier::ReferablePaths::CATEGORY, category);
}

std::string ConnectedDataElement::getDescription() const
{
  return this->getProxyValue(qualifier::ReferablePaths::DESCRIPTION);
}

void ConnectedDataElement::setDescription(const std::string & description)
{
  this->setProxyValue(qualifier::ReferablePaths::DESCRIPTION, description);
}

basyx::any ConnectedDataElement::getParent() const
{
  return this->getProxyCollection(reference::paths::PARENTS);
}

void ConnectedDataElement::setParent(const basyx::any & parent_reference)
{
  this->setProxyValue(reference::paths::PARENTS, parent_reference);
}

std::string ConnectedDataElement::getId() const
{
  return this->getIdWithLocalCheck();
}

void ConnectedDataElement::setId(const std::string & id)
{
  this->setIdWithLocalCheck(id);
}

basyx::objectCollection_t ConnectedDataElement::getQualifier() const
{
  return this->getProxyCollection(reference::paths::QUALIFIERS);
}

void ConnectedDataElement::setQualifier(const basyx::objectCollection_t & qualifiers)
{
  this->setProxyValue(reference::paths::QUALIFIERS, qualifiers);
}

basyx::any ConnectedDataElement::getSemanticId() const
{
  return this->getProxyValue(reference::paths::SEMANTICIDS);
}

void ConnectedDataElement::setSemanticID(const basyx::any & semantic_identifier)
{
  this->setProxyValue(reference::paths::SEMANTICIDS, semantic_identifier);
}

std::string ConnectedDataElement::getHasKindReference() const
{
  return this->getProxyValue(qualifier::haskind::Paths::KIND);
}

void ConnectedDataElement::setHasKindReference(const std::string & kind)
{
  this->setProxyValue(qualifier::haskind::Paths::KIND, kind);
}

void ConnectedDataElement::setProxyValue(const std::string & path, const basyx::any value) const
{
  this->getProxy()->updateElementValue(path, value);
}

std::string ConnectedDataElement::getProxyValue(const std::string & path) const
{
  auto value = getProxy()->readElementValue(path);
  return value.Get<std::string>();
}

basyx::objectCollection_t ConnectedDataElement::getProxyCollection(const std::string & path) const
{
  auto value = this->getProxy()->readElementValue(path);
  return value.Get<basyx::objectCollection_t>();
}

std::string ConnectedDataElement::getIdWithLocalCheck() const
{
  basyx::any localId = this->getLocalValue(qualifier::ReferablePaths::IDSHORT);
  if ( not localId.IsNull() )
  {
    return localId.Get<std::string>();
  }
  return this->getIdShort();
}

void ConnectedDataElement::setIdWithLocalCheck(const std::string & id)
{
  auto localId = this->getLocalValue(qualifier::ReferablePaths::IDSHORT);
  if ( not localId.IsNull() )
  {
    this->updateLocalValue(qualifier::ReferablePaths::IDSHORT, id);
    return;
  }
  this->setProxyValue(qualifier::ReferablePaths::IDSHORT, id);
}

}
}
}
